### PR TITLE
DDPB-3671: Refactor email sending to be testable and remove controller tests

### PR DESCRIPTION
**Pull request by acsauk**
## Purpose
As part of the Symfony upgrade, we're working our way through removing instances in the app where we're getting services directly from the DI container in code (`$container->get('<SERVICE ID>'`) or where we're setting services in code or tests (`$container->set('SERVICE ID')`) to ensure we're ready to go private by default and benefit from speed boosts of the upgrade. The main area where we were doing this was in controller tests around sending emails so this PR  aims to move as much logic out of the controllers and into an event-based system to keep business logic together and aid testing.

Fixes DDPB-3671

## Approach
In order to make email sending testable and consistent, I've started to emit events when common transactions happen within the app (ReportSubmitted, UserEdited etc). On the most part this happens when we do something with an entity so it made sense to wrap up calls to our internal API layer in classes and emit the event after a successful transaction. This opens up the use of event subscribers to listen out for these events and isolate and test the business logic. While this mainly focuses on email sending and event logging for now we could start to adopt this approach more widely through the app to keep as much logic out of controllers as possible.

## Learning
Testing controllers in our app is still a PITA. I'd hoped to use a mock event dispatcher that collects the events dispatched to look at later but it looks like some of our `RestClient` functionality is still getting in the way and making API calls in the background, potentially linked to some existing event listeners. I've removed the controller tests for now but we should look to find a better way of testing controllers once the upgrade to 4/5 is completed.

## Checklist
- [x] I have performed a self-review of my own code
- [ ] I have updated documentation (Confluence/ADR/tech debt doc) where relevant
- [x] I have added tests to prove my work
- [ ] The product team have approved these changes

### CODE REVIEW

**Review by coderste — APPROVED** [review]
LGTM 👍🏻 left a few minor comments but they don't really need looking at so I've approved it :)

File: api/src/AppBundle/Entity/User.php [hunk 1]
```diff
@@ -1278,4 +1278,17 @@ public function hasReports()
 
         return false;
     }
+
+    public function isCoDeputyWith(User $coDeputy)
```
[the remark this answers is not in the document]
Comment (acsauk) [reply api/src/AppBundle/Entity/User.php:1282]: Yup - I'll get that added 👍

File: client/src/AppBundle/Controller/Org/TeamController.php [hunk 2]
```diff
@@ -168,9 +146,12 @@ public function editAction(Request $request, $id)
 
         $validationGroups = $team->canAddAdmin() ? ['user_details_org', 'org_team_role_name'] : ['user_details_org'];
 
-        $form = $this->createForm(FormDir\Org\TeamMemberAccountType::class, $user, ['team' => $team, 'loggedInUser' => $this->getUser(), 'targetUser' => $user, 'validation_groups' => $validationGroups
+        $form = $this->createForm(
+            FormDir\Org\TeamMemberAccountType::class,
+            $user,
+            ['team' => $team, 'loggedInUser' => $this->getUser(), 'targetUser' => $user, 'validation_groups' => $validationGroups
```
[the remark this answers is not in the document]
Comment (acsauk) [reply client/src/AppBundle/Controller/Org/TeamController.php:152]: No, I agree - I'd always take readability over line count. Will fix up.

**Review by coderste — APPROVED** [review]

**Review by Gugandeep — APPROVED** [review]
Great work! Looks good to me.

**Pull request merged**